### ISSUE TITLE
Significant ingest performance regression while handling parsing dynamically added objects

### ISSUE THREAD

**Issue by benwtrent** [elastic/elasticsearch#103011]
### Elasticsearch Version

8.11+

### Java Version

_bundled_

### OS Version

any

### Problem Description

While indexing documents with dynamically added fields, if multiple of those fields are objects are are deeper than 3-5 objects, performance is degraded significantly.

This bug was introduced in: https://github.com/elastic/elasticsearch/pull/90674

### Steps to Reproduce

I built a micro-benchmark to replicate this particular bug. 

But, any dynamic object parsing will cause this.

To mitigate the issue, statically mapped fields for objects is required. 

### Logs (if relevant)

Attached are two JFR files indicating the difference between when the bug was introduce vs. not.
[Archive 2.zip](https://github.com/elastic/elasticsearch/files/13569248/Archive.2.zip)

**Comment by benwtrent** [comment]
I have confirmed this is still not fixed. https://github.com/elastic/elasticsearch/pull/103047 improved parsing by 10x, but its still about 10x slower than before. 

So now, in 8.11.3, instead of 100x slower, parsing is only 10x slower. Digging into the cause to see what other options there are.

I adjusted the benchmark to add some variety to the document "nested ness".

Before this fix:
 - `150584.973 ms/op`

After this fix:
 - `11237.001 ms/op`

So, an improvement of over a magnitude. But, previous to 8.11:
 - `692.411 ms/op`

So, still about a magnitude slower.
